core/synch.c

```diff
@@ -1749,6 +1749,12 @@ translate_from_synchall_to_dispatch(thread_record_t *tr, thread_synch_state_t sy
                 arch_mcontext_reset_stolen_reg(dcontext, mc);
             }
         });
+        IF_AARCHXX({
+            set_stolen_reg_val(mc, (reg_t)os_get_dr_tls_base(dcontext));
```

Comment:
I wonder if this should be fixed in a separate PR since it is a separate bug.

Reply:
Without this part, this PR would be incomplete as the large app crashes without the properly set up stolen reg. 

If this part is a step in the right direction and you feel there's more work to be done on stolen reg handling here, I can add the TODO with #4495 and #4460 both.

Reply:
The separate PR would go in before this one of course.  But this is a direct bug fix of arm synchall, which is a separate thing from the new callback.  IHMO the arm synchall bug should have its own issue filed, and its own separate fix.

Reply:
I don't think this is related to #4460: there is no evidence that it is, right?  There's no synchall in the #4460 scenario.  I've also put this fix in a local tree and #4460 still reproduces.

Maybe this synchall stolen reg restore could be bundled into #4495 and we leave it open for the further fix for synchall and for signals, rather than a separate issue.

Reply:
> The separate PR would go in before this one of course. But this is a direct bug fix of arm synchall, which is a separate thing from the new callback. IHMO the arm synchall bug should have its own issue filed, and its own separate fix.

Oh that makes sense. I can file a separate issue and PR for this.

Earlier, I thought you were suggesting leaving out the stolen reg restore completely for now, perhaps because we haven't yet figured out the stolen reg bug completely (pending #4460 and #4495)

> Maybe this synchall stolen reg restore could be bundled into #4495 and we leave it open for the further fix for synchall and for signals, rather than a separate issue.

I think it's better to fix the known part in a separate PR, as suggested above.

Reply:
Separated out the stolen reg changes in PR #4498.

Reply:
> > Maybe this synchall stolen reg restore could be bundled into #4495 and we leave it open for the further fix for synchall and for signals, rather than a separate issue.
> 
> I think it's better to fix the known part in a separate PR, as suggested above.

FTR #4495 is an issue: I meant bundling the two synchall stolen reg problems into the same issue (fixed with separate PR's).  But a separate issue is fine too.

Reply:
I ended up filing a separate issue for the two. #4497 (loading the stolen reg before reset) will need to be resolved more immediately (as it's actually causing some crashes) vs #4495 (saving the translated stolen reg value) needs more discussion and may need more time.